File: 백준/Silver/2178. 미로 탐색/미로 탐색.swift
import Foundation

var input = readLine()!.split(separator: " ").map { Int(String($0))! }

let dx = [0, 0, -1, 1]
let dy = [-1, 1, 0, 0] 

var arr = [[Int]]()
var visited = Array(repeating: Array(repeating: false, count: input[1]), count: input[0])
var distance = Array(repeating: Array(repeating: 0, count: input[1]), count: input[0])

for _ in 0..<input[0] {
    var temp = Array(readLine()!).map { Int(String($0))! }
    arr.append(temp)
}

func bfs() {
    distance[0][0] = 1
    visited[0][0] = true
    
    var queue: [[Int]] = [[0,0]]
    
    while !queue.isEmpty {
        let pop = queue.removeFirst()
        
        let x = pop[0]
        let y = pop[1]
        
        for i in 0..<4 {
            let nx = x + dx[i]
            let ny = y + dy[i]
            
            if nx >= 0 && nx < input[0] && ny >= 0 && ny < input[1] {
                if !visited[nx][ny] && arr[nx][ny] == 1 {
                    distance[nx][ny] = distance[x][y] + 1
                    
                    visited[nx][ny] = true
                    queue.append([nx, ny])
                }
            }
        }
    }
}

bfs()
print(distance[input[0]-1][input[1]-1])

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int M = sc.nextInt();
        
        int[][] arr = new int[N][M];
        int[][] dist = new int[N][M]; // 각 칸에 가기위한 거리 측정용
        Queue<int[]> q = new LinkedList<>();

        int[] dx = {-1,0,1,0};
        int[] dy = {0,1,0,-1};
        
        for(int i=0; i<N; i++) {
            int j = 0;
            char[] input = sc.next().toCharArray();
            for(char c : input) {
                arr[i][j] = Character.getNumericValue(c);
                j++;
            }
        }

        // BFS 탐색 시작
        dist[0][0] = 1;
        q.offer(new int[] {0,0});
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            for(int d=0; d<4; d++) {
                int nx = cur[0] + dx[d];
                int ny = cur[1] + dy[d];
                if(nx > -1 && nx < N && ny > -1 && ny < M && dist[nx][ny] == 0 && arr[nx][ny] == 1) {
                    q.offer(new int[] {nx,ny});
                    dist[nx][ny] = dist[cur[0]][cur[1]] + 1;
                }
            }
        }
        System.out.println(dist[N-1][M-1]);
    }
}